dev/tools/localization/gen_l10n.dart

```diff
@@ -473,13 +470,94 @@ Future<void> main(List<String> arguments) async {
+  LocalizationsGenerator(this._fs);
+
+  final file.FileSystem _fs;
+
+  /// The reference to the project's l10n directory.
+  ///
+  /// It is assumed that all input files (ie. [templateArbFile], arb files
+  /// for translated messages) and output files (ie. The localizations
+  /// [outputFile], `messages_<locale>.dart` and `messages_all.dart`)
+  /// will reside here.
+  Directory l10nDirectory;
+
+  /// The reference to the template arb file.
+  ///
+  /// The messages and its metadata will be used to generate the
+  /// `messages_<locale>.dart` and `messages_all.dart` files.
+  File templateArbFile;
+
+  /// The reference to the Dart file will contain the localizations and
+  /// localizations delegate classes.
+  File outputFile;
+
+  /// The class name to be used for the localizations class in [outputFile].
+  String className;
+
+  void setL10nDirectory(String arbPathString) {
+    if (arbPathString == null)
+      throw L10nException('Input string cannot be null');
+    l10nDirectory = _fs.directory(arbPathString);
+    if (!l10nDirectory.existsSync())
+      throw FileSystemException(
+        "The 'arb-dir' directory, $l10nDirectory, does not exist.\n"
+        'Make sure that the correct path was provided.'
+      );
+    final String l10nDirectoryStatModeString = l10nDirectory.statSync().modeString();
+    if (!_isDirectoryReadableAndWritable(l10nDirectoryStatModeString))
+      throw FileSystemException(
+        "The 'arb-dir' directory, $l10nDirectory, doesn't allow reading and writing.\n"
+        'Please ensure that the user has read and write permissions.'
+      );
+  }
+
+  void setTemplateArbFile(String templateArbFileName) {
+    if (templateArbFileName == null)
+      throw L10nException('Input string cannot be null');
+    if (l10nDirectory == null)
+      throw L10nException('l10nDirectory cannot be null when setting template arb file');
+
+    templateArbFile = _fs.file(path.join(l10nDirectory.path, templateArbFileName));
+    final String templateArbFileStatModeString = templateArbFile.statSync().modeString();
+    if (templateArbFileStatModeString[0] == '-' && templateArbFileStatModeString[3] == '-')
+      throw FileSystemException(
+        "The 'template-arb-file', $templateArbFile, is not readable.\n"
+        'Please ensure that the user has read permissions.'
+      );
+  }
+
+  void setOutputFile(String outputFileString) {
+    if (outputFileString == null)
+      throw L10nException('Input string cannot be null');
+    outputFile = _fs.file(path.join(l10nDirectory.path, outputFileString));
+  }
+
+  void setClassName(String classNameString) {
+    if (classNameString == null)
+      throw L10nException('Input string cannot be null');
+    if (!_isValidClassName(classNameString))
+      throw L10nException(
+        "The 'output-class', $classNameString, is not a valid Dart class name.\n"
+      );
+    className = classNameString;
+  }
+}
+
+class L10nException implements Exception {
```

Comment:
I'm not quite sure how to find existing exceptions that make sense for the error cases I'm writing. I created this as a placeholder for now, but is there a sane way to figure out if an existing exception class is better for my use cases? Or is it normal to create your own?

Reply:
I would either create your own or re-use Exception.